File: scraper.js
const {
  dedupe,
  extractContactDetailsFromText,
  extractContactLinksFromMarkdown,
  parseBingSearchMarkdown,
} = require("./scraper-helpers");

const SEARCH_USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36";
const LINKEDIN_PRIORITY_REGIONS = [
  "India",
  "Bengaluru",
  "Bangalore",
  "Mumbai",
  "Hyderabad",
  "Pune",
  "Chennai",
  "Delhi",
  "Gurugram",
  "Noida",
  "Kolkata",
];
const LINKEDIN_OTHER_REGIONS = [
  "United States",
  "USA",
  "United Kingdom",
  "UK",
  "Canada",
  "Singapore",
  "UAE",
  "Dubai",
  "Middle East",
];
const HR_TERMS = ["HR", "recruiter", "human resources", "talent acquisition", "hiring"];

function linkedinSearchVariants(keyword) {
  const base = keyword.trim();
  const queries = [];
  const seen = new Set();

  const add = (query) => {
    const normalized = query.trim();
    if (!normalized || seen.has(normalized)) return;
    seen.add(normalized);
    queries.push(normalized);
  };

  for (const region of LINKEDIN_PRIORITY_REGIONS) {
    for (const hrTerm of HR_TERMS) {
      add(`site:linkedin.com/posts "${base}" "${region}" "${hrTerm}" email`);
      add(`site:linkedin.com/feed/update "${base}" "${region}" "${hrTerm}" email`);
      add(`site:linkedin.com/company "${base}" "${region}" "${hrTerm}" email`);
      add(`site:linkedin.com/in "${base}" "${region}" "${hrTerm}" email`);
    }
  }

  for (const region of LINKEDIN_OTHER_REGIONS) {
    for (const hrTerm of HR_TERMS) {
      add(`site:linkedin.com/posts "${base}" "${region}" "${hrTerm}" email`);
      add(`site:linkedin.com/feed/update "${base}" "${region}" "${hrTerm}" email`);
    }
  }

  add(`site:linkedin.com/posts "${base}" hiring email`);
  add(`site:linkedin.com/feed/update "${base}" hiring email`);
  add(`site:linkedin.com/jobs "${base}" email`);
  add(`site:linkedin.com/company "${base}" hiring email`);

  return queries;
}

const SEARCH_VARIANTS = (keyword, source = "public") => {
  if (source === "linkedin") {
    return linkedinSearchVariants(keyword);
  }

  return [keyword, `${keyword} contact`, `${keyword} email`, `${keyword} phone`];
};

function proxyUrl(url) {
  return `https://r.jina.ai/http://${url.replace(/^https?:\/\//i, "")}`;
}

async function fetchProxyText(url, timeoutMs = 20000) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(proxyUrl(url), {
      signal: controller.signal,
      headers: {
        "User-Agent": SEARCH_USER_AGENT,
        Accept: "text/plain, text/markdown;q=0.9, */*;q=0.1",
      },
    });

    if (!response.ok) {
      throw new Error(`Proxy fetch failed for ${url} (${response.status})`);
    }

    return response.text();
  } finally {
    clearTimeout(timer);
  }
}

function sameOrigin(urlA, urlB) {
  try {
    return new URL(urlA).origin === new URL(urlB).origin;
  } catch {
    return false;
  }
}

function normalizeWebsiteUrl(url) {
  try {
    const parsed = new URL(url);
    parsed.hash = "";
    return parsed.href;
  } catch {
    return "";
  }
}

function searchFreshnessParam(maxAgeHours) {
  const hours = Number(maxAgeHours) || 0;
  if (hours <= 0) return "";
  if (hours <= 24) return "&freshness=Day";
  if (hours <= 168) return "&freshness=Week";
  return "&freshness=Month";
}

function inferLinkedInRegion(text) {
  const value = String(text || "").toLowerCase();
  if (!value) return "";
  for (const region of LINKEDIN_PRIORITY_REGIONS) {
    if (value.includes(region.toLowerCase())) return region;
  }
  for (const region of LINKEDIN_OTHER_REGIONS) {
    if (value.includes(region.toLowerCase())) return region;
  }
  return "";
}

function inferExperience(text) {
  const value = String(text || "");
  const rangeMatch = value.match(/\b(\d{1,2})\s*(?:-|-|to)\s*(\d{1,2})\s*(?:years?|yrs?)\b/i);
  if (rangeMatch) return `${rangeMatch[1]}-${rangeMatch[2]} years`;

  const plusMatch = value.match(/\b(\d{1,2})\s*\+\s*(?:years?|yrs?)\b/i);
  if (plusMatch) return `${plusMatch[1]}+ years`;

  const experienceMatch = value.match(/\b(\d{1,2})\s*(?:\+)?\s*(?:years?|yrs?)\s*(?:of\s+)?experience\b/i);
  if (experienceMatch) return `${experienceMatch[1]}+ years`;

  const numericMatch = value.match(/\b(\d{1,2})\s*(?:years?|yrs?)\b/i);
  if (numericMatch) return `${numericMatch[1]} years`;

  if (/\bfresher\b/i.test(value)) return "Fresher";
  return "";
}

function scoreLinkedInResult(result) {
  const text = `${result.title || ""} ${result.snippet || ""} ${result.query || ""}`.toLowerCase();
  let score = 0;
  if (LINKEDIN_PRIORITY_REGIONS.some((region) => text.includes(region.toLowerCase()))) score += 30;
  if (LINKEDIN_OTHER_REGIONS.some((region) => text.includes(region.toLowerCase()))) score += 10;
  if (HR_TERMS.some((term) => text.includes(term.toLowerCase()))) score += 25;
  if (/\bemail\b/i.test(text)) score += 8;
  if (/\bhiring\b/i.test(text)) score += 8;
  if (/\brecruiter\b/i.test(text)) score += 10;
  if (/\bhuman resources\b/i.test(text)) score += 10;
  return score;
}

async function collectSearchResults(keyword, maxSearchPages, source = "public") {
  const results = [];
  const seen = new Set();
  const variants = SEARCH_VARIANTS(keyword, source);
  const pagesPerVariant = Math.max(1, Number(maxSearchPages) || 1);
  let pagesChecked = 0;

  for (const query of variants) {
    for (let pageIndex = 0; pageIndex < pagesPerVariant; pageIndex += 1) {
      const searchUrl =
        pageIndex === 0
          ? `https://www.bing.com/search?q=${encodeURIComponent(query)}`
          : `https://www.bing.com/search?q=${encodeURIComponent(query)}&first=${pageIndex * 10}`;
      const markdown = await fetchProxyText(searchUrl, 20000);
      const parsed = parseBingSearchMarkdown(markdown, query, pageIndex + 1);
      pagesChecked += 1;

      for (const result of parsed) {
        const website = normalizeWebsiteUrl(result.url);
        if (!website || seen.has(website)) continue;
        seen.add(website);
        results.push({ ...result, url: website });
      }
    }
  }

  return { results, pagesChecked };
}

function extractExternalLinksFromMarkdown(markdown, baseUrl, limit = 3) {
  const links = [];
  const seen = new Set();
  const linkRegex = /\[([^\]]*)\]\(([^)]+)\)/g;
  let match;

  while ((match = linkRegex.exec(String(markdown || "")))) {
    try {
      const absolute = new URL(match[2].trim(), baseUrl).href;
      const parsed = new URL(absolute);
      if (!/^https?:$/i.test(parsed.protocol)) continue;
      if (parsed.hostname.toLowerCase().includes("linkedin.com")) continue;
      parsed.hash = "";
      parsed.search = "";
      const normalized = parsed.href;
      if (seen.has(normalized)) continue;
      seen.add(normalized);
      links.push(normalized);
      if (links.length >= limit) break;
    } catch {
      // Ignore invalid URLs.
    }
  }

  return links;
}

async function fetchSiteDetails(url, emailFragment, source = "public") {
  const visited = [];
  const details = [];
  const rootMarkdown = await fetchProxyText(url, 20000);
  visited.push(url);

  const rootContactLinks =
    source === "linkedin"
      ? extractExternalLinksFromMarkdown(rootMarkdown, url, 2)
      : extractContactLinksFromMarkdown(rootMarkdown, url).slice(0, 3);
  const pagesToVisit = dedupe([url, ...rootContactLinks]);

  for (const pageUrl of pagesToVisit) {
    if (visited.includes(pageUrl)) {
      const { emails, phones } = extractContactDetailsFromText(rootMarkdown, rootMarkdown, emailFragment);
      if (emails.length || phones.length) {
        details.push({
          pageUrl,
          emails,
          phones,
          sourceText: rootMarkdown,
        });
      }
      continue;
    }

    try {
      const pageText = await fetchProxyText(pageUrl, 15000);
      visited.push(pageUrl);
      const { emails, phones } = extractContactDetailsFromText(pageText, pageText, emailFragment);
      if (emails.length || phones.length) {
        details.push({
          pageUrl,
          emails,
          phones,
          sourceText: pageText,
        });
      }
    } catch {
      // Skip pages the proxy cannot fetch.
    }
  }

  return { visited, details };
}

async function scrapeLeads({
  keyword,
  emailFragment = "",
  maxSearchPages = 1,
  maxWebsites = 3,
  source = "public",
  maxAgeHours = 0,
}) {
  const variants = SEARCH_VARIANTS(keyword, source);
  const searchResults = [];
  const seenSearchResults = new Set();
  let searchPagesChecked = 0;

  for (const query of variants) {
    for (let pageIndex = 0; pageIndex < Math.max(1, Number(maxSearchPages) || 1); pageIndex += 1) {
      const searchUrl =
        pageIndex === 0
          ? `https://www.bing.com/search?q=${encodeURIComponent(query)}${searchFreshnessParam(maxAgeHours)}`
          : `https://www.bing.com/search?q=${encodeURIComponent(query)}&first=${pageIndex * 10}${searchFreshnessParam(maxAgeHours)}`;
      const markdown = await fetchProxyText(searchUrl, 20000);
      const parsed = parseBingSearchMarkdown(markdown, query, pageIndex + 1);
      searchPagesChecked += 1;

      for (const result of parsed) {
        const website = normalizeWebsiteUrl(result.url);
        if (!website || seenSearchResults.has(website)) continue;
        try {
          const origin = new URL(website);
          if (origin.hostname.toLowerCase().includes("bing.com")) continue;
          if (source === "linkedin" && !origin.hostname.toLowerCase().includes("linkedin.com")) continue;
          seenSearchResults.add(website);
          searchResults.push({ ...result, url: website });
        } catch {
          // Skip invalid results.
        }
      }
    }
  }

  if (source === "linkedin") {
    searchResults.sort((left, right) => scoreLinkedInResult(right) - scoreLinkedInResult(left));
  }

  const results = [];
  const seenSites = new Set();
  const seenContacts = new Set();
  let websitesChecked = 0;
  let pagesChecked = 0;

  for (const searchResult of searchResults.slice(0, maxWebsites * 3)) {
    if (results.length >= maxWebsites) break;
    const website = searchResult.url;

    let origin;
    try {
      origin = new URL(website).origin;
    } catch {
      continue;
    }

    if (seenSites.has(origin)) continue;

    seenSites.add(origin);
    websitesChecked += 1;

    try {
      const { visited, details } = await fetchSiteDetails(website, emailFragment, source);
      pagesChecked += visited.length;

      for (const detail of details) {
        for (const email of detail.emails) {
          for (const phone of detail.phones.length ? detail.phones : [""]) {
            const signature = `${origin}|${email}|${phone}|${detail.pageUrl}`;
            if (seenContacts.has(signature)) continue;
            seenContacts.add(signature);
            results.push({
              keyword,
              website: origin,
              title: searchResult.title || "",
              email,
              phone,
              region: inferLinkedInRegion(`${searchResult.title || ""} ${searchResult.snippet || ""}`) || "",
              experience: inferExperience(`${searchResult.title || ""} ${searchResult.snippet || ""} ${detail.sourceText || ""}`) || "",
              sourcePage: detail.pageUrl,
              searchQuery: searchResult.query || keyword,
              searchTitle: searchResult.title || "",
              searchSnippet: searchResult.snippet || "",
              ageHours: maxAgeHours || "",
            });
          }
        }
      }

      if (!results.length) {
        const { emails, phones } = extractContactDetailsFromText(
          details[0]?.sourceText || "",
          details[0]?.sourceText || "",
          emailFragment
        );
        for (const email of emails.slice(0, 3)) {
          const phone = phones[0] || "";
          const signature = `${origin}|${email}|${phone}|${website}`;
          if (seenContacts.has(signature)) continue;
          seenContacts.add(signature);
          results.push({
            keyword,
            website: origin,
            title: searchResult.title || "",
            email,
            phone,
            region: inferLinkedInRegion(`${searchResult.title || ""} ${searchResult.snippet || ""}`) || "",
            experience: inferExperience(`${searchResult.title || ""} ${searchResult.snippet || ""} ${details[0]?.sourceText || ""}`) || "",
            sourcePage: website,
            searchQuery: searchResult.query || keyword,
            searchTitle: searchResult.title || "",
            searchSnippet: searchResult.snippet || "",
            ageHours: maxAgeHours || "",
          });
        }
      }
    } catch {
      // Skip sites that fail to fetch.
    }
  }

  return {
    keyword,
    emailFragment,
    source,
    searchPagesChecked,
    websitesChecked,
    pagesChecked,
    searchSource: source === "linkedin" ? "bing-linkedin" : "bing",
    searchNotice: "",
    results,
  };
}

module.exports = {
  scrapeLeads,
  collectSearchResults,
  fetchSiteDetails,
};

File: scraper.py
import base64
import json
import re
import sys
import time
import urllib.parse
import urllib.request
import urllib.error
from html import unescape


USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
)

EMAIL_RE = re.compile(r"[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}", re.I)
PHONE_RE = re.compile(
    r"(?:\+?\d{1,3}[\s.\-]?)?(?:\(?\d{2,4}\)?[\s.\-]?)?\d{3,4}[\s.\-]?\d{3,4}(?:[\s.\-]?\d{2,4})?"
)


def _clean_url(url: str) -> str:
    try:
        parsed = urllib.parse.urlsplit(url)
        parsed = parsed._replace(fragment="")
        return urllib.parse.urlunsplit(parsed)
    except Exception:
        return url.strip()


def _proxy_url(url: str) -> str:
    url = url.strip()
    if url.startswith("https://"):
        return "https://r.jina.ai/http://" + url[len("https://") :]
    if url.startswith("http://"):
        return "https://r.jina.ai/http://" + url[len("http://") :]
    return "https://r.jina.ai/http://" + url


def fetch_proxy_text(url: str, timeout: int = 12) -> str:
    last_error = None
    for attempt in range(4):
        req = urllib.request.Request(
            _proxy_url(url),
            headers={
                "User-Agent": USER_AGENT,
                "Accept": "text/plain, text/markdown;q=0.9, */*;q=0.1",
            },
        )
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return resp.read().decode("utf-8", errors="replace")
        except urllib.error.HTTPError as exc:
            last_error = exc
            if exc.code != 429:
                raise
            time.sleep(0.6 * (attempt + 1))
        except urllib.error.URLError as exc:
            last_error = exc
            if attempt == 3:
                raise
            time.sleep(0.4 * (attempt + 1))

    raise last_error if last_error else RuntimeError("Proxy fetch failed")


def decode_bing_redirect(url: str) -> str:
    try:
        parsed = urllib.parse.urlsplit(url)
        params = urllib.parse.parse_qs(parsed.query)
        encoded = params.get("u", [""])[0]
        if encoded.startswith("a1"):
            payload = encoded[2:]
            payload += "=" * ((4 - len(payload) % 4) % 4)
            decoded = base64.b64decode(payload).decode("utf-8", errors="replace")
            if decoded.startswith(("http://", "https://")):
                return _clean_url(decoded)
    except Exception:
        pass
    return _clean_url(url)


def parse_bing_search_markdown(markdown: str, query: str, page_index: int):
    results = []
    current = None

    for raw_line in markdown.splitlines():
        line = raw_line.rstrip()
        match = re.match(r"^\s*(\d+)\.\s+##\s+\[(.+?)\]\((.+?)\)\s*$", line)
        if match:
            if current:
                results.append(current)
            current = {
                "rank": int(match.group(1)),
                "title": unescape(re.sub(r"\*\*", "", match.group(2))).strip(),
                "url": decode_bing_redirect(match.group(3).strip()),
                "snippet": "",
                "query": query,
                "pageIndex": page_index,
            }
            continue

        if not current:
            continue

        if not line.strip():
            if current.get("snippet"):
                results.append(current)
                current = None
            continue

        snippet_line = re.sub(r"^\s*[-*]\s+", "", re.sub(r"\*\*", "", line)).strip()
        if current["snippet"]:
            current["snippet"] += " " + snippet_line
        else:
            current["snippet"] = snippet_line

    if current:
        results.append(current)

    return [item for item in results if item["url"].startswith(("http://", "https://"))]


def extract_emails(text: str, fragment: str = ""):
    values = []
    for email in EMAIL_RE.findall(text or ""):
        email = email.strip("<>()[]{}\"'` ")
        if fragment and fragment.lower() not in email.lower():
            continue
        domain = email.split("@", 1)[-1]
        if re.search(r"\.(?:jpg|jpeg|png|gif|webp|svg|ico)$", domain, re.I):
            continue
        values.append(email)
    return sorted(set(values))


def extract_phones(text: str):
    phones = []
    for phone in PHONE_RE.findall(text or ""):
        phone = re.sub(r"\s+", " ", phone.replace("\xa0", " ")).strip()
        digits = re.sub(r"\D", "", phone)
        if len(digits) < 10 or len(digits) > 15:
            continue
        if not re.search(r"[+().\-\s]", phone):
            continue
        phones.append(phone)
    return sorted(set(phones))


def extract_contact_links(markdown: str, base_url: str):
    keywords = (
        "contact",
        "contact-us",
        "contactus",
        "about",
        "team",
        "support",
        "reach",
        "locations",
        "location",
        "office",
        "impressum",
    )
    links = []
    seen = set()
    for label, href in re.findall(r"\[([^\]]*)\]\(([^)]+)\)", markdown or ""):
        candidate = f"{label} {href}".lower()
        if not any(word in candidate for word in keywords):
            continue
        try:
            absolute = urllib.parse.urljoin(base_url, href.strip())
            if absolute not in seen:
                seen.add(absolute)
                links.append(absolute)
        except Exception:
            continue
    return links


def extract_external_links(markdown: str, base_url: str, limit: int = 3):
    links = []
    seen = set()
    for _label, href in re.findall(r"\[([^\]]*)\]\(([^)]+)\)", markdown or ""):
        try:
            absolute = urllib.parse.urljoin(base_url, href.strip())
            parsed = urllib.parse.urlsplit(absolute)
            if parsed.scheme not in ("http", "https"):
                continue
            if "linkedin.com" in parsed.netloc.lower():
                continue
            cleaned = urllib.parse.urlunsplit((parsed.scheme, parsed.netloc, parsed.path, "", ""))
            if cleaned in seen:
                continue
            seen.add(cleaned)
            links.append(cleaned)
            if len(links) >= limit:
                break
        except Exception:
            continue
    return links


def search_queries(keyword: str, source: str = "public"):
    keyword = keyword.strip()
    if source == "linkedin":
        return [
            f'site:linkedin.com/posts "{keyword}" hiring email',
            f'site:linkedin.com/feed/update "{keyword}" hiring email',
            f'site:linkedin.com/jobs "{keyword}" email',
            f'site:linkedin.com/company "{keyword}" hiring email',
        ]

    return [
        keyword,
        f"{keyword} contact",
        f"{keyword} email",
    ]


def collect_search_results(keyword: str, max_search_pages: int, source: str = "public"):
    results = []
    seen = set()
    pages_checked = 0

    for query in search_queries(keyword, source):
        for page_index in range(max_search_pages):
            url = f"https://www.bing.com/search?q={urllib.parse.quote(query)}"
            if page_index > 0:
                url += f"&first={page_index * 10}"

            try:
                markdown = fetch_proxy_text(url, timeout=10)
            except urllib.error.HTTPError as exc:
                if exc.code == 429:
                    yield {
                        "type": "search_rate_limited",
                        "searchPagesChecked": pages_checked,
                        "searchQuery": query,
                        "pageIndex": page_index + 1,
                    }
                    continue
                raise
            parsed = parse_bing_search_markdown(markdown, query, page_index + 1)
            pages_checked += 1
            yield {
                "type": "search_progress",
                "searchPagesChecked": pages_checked,
                "searchQuery": query,
                "pageIndex": page_index + 1,
                "resultCount": len(results),
            }
            for item in parsed:
                clean_url = _clean_url(item["url"])
                try:
                    origin = urllib.parse.urlsplit(clean_url)
                    if origin.scheme not in ("http", "https"):
                        continue
                    if "bing.com" in origin.netloc.lower():
                        continue
                    if source == "linkedin" and "linkedin.com" not in origin.netloc.lower():
                        continue
                    normalized = urllib.parse.urlunsplit((origin.scheme, origin.netloc, origin.path, "", ""))
                    if normalized in seen:
                        continue
                    seen.add(normalized)
                    results.append({**item, "url": normalized})
                except Exception:
                    continue

    yield {
        "type": "search_done",
        "results": results,
        "pagesChecked": pages_checked,
    }


def fetch_site_details(url: str, fragment: str = "", source: str = "public"):
    visited = []
    details = []
    try:
        root_text = fetch_proxy_text(url, timeout=10)
    except urllib.error.HTTPError as exc:
        if exc.code == 429:
            return {"visited": visited, "details": [], "notice": "rate_limited"}
        raise
    visited.append(url)
    if source == "linkedin":
        candidate_pages = [url] + extract_external_links(root_text, url, limit=2)
    else:
        candidate_pages = [url] + extract_contact_links(root_text, url)[:3]

    for page_url in candidate_pages:
        try:
            if page_url == url:
                page_text = root_text
            else:
                page_text = fetch_proxy_text(page_url, timeout=8)
            if page_url != url:
                visited.append(page_url)
            emails = extract_emails(page_text, fragment)
            phones = extract_phones(page_text)
            if emails or phones:
                details.append(
                    {
                        "pageUrl": page_url,
                        "emails": emails,
                        "phones": phones,
                        "sourceText": page_text,
                    }
                )
        except Exception:
            continue

    return {"visited": visited, "details": details}


def scrape_leads(
    keyword: str,
    email_fragment: str = "",
    max_search_pages: int = 1,
    max_websites: int = 3,
    source: str = "public",
):
    search_results = []
    search_pages_checked = 0
    for event in collect_search_results(keyword, max_search_pages, source):
        if isinstance(event, dict) and event.get("type") == "search_progress":
            search_pages_checked = event.get("searchPagesChecked", search_pages_checked)
            continue
        if isinstance(event, dict) and event.get("type") == "search_done":
            search_results = event["results"]
            search_pages_checked = event.get("pagesChecked", search_pages_checked)

    results = []
    seen_sites = set()
    seen_contacts = set()
    websites_checked = 0
    pages_checked = 0
    search_notice = ""

    for search_result in search_results[: max_websites * 3]:
        if len(results) >= max_websites:
            break
        website = search_result["url"]
        try:
            origin = urllib.parse.urlsplit(website)
            site_key = urllib.parse.urlunsplit((origin.scheme, origin.netloc, "", "", ""))
        except Exception:
            continue
        if site_key in seen_sites:
            continue
        seen_sites.add(site_key)
        websites_checked += 1

        try:
            details = fetch_site_details(website, email_fragment, source)
            pages_checked += len(details["visited"])
            if details.get("notice") == "rate_limited":
                search_notice = "Some pages were rate limited, so results may be incomplete."
            yield {
                "type": "site_progress",
                "websitesChecked": websites_checked,
                "pagesChecked": pages_checked,
                "resultsCount": len(results),
                "currentWebsite": site_key,
            }
            for detail in details["details"]:
                emails = detail["emails"] or [""]
                phones = detail["phones"] or [""]
                for email in emails:
                    for phone in phones:
                        signature = f"{site_key}|{email}|{phone}|{detail['pageUrl']}"
                        if signature in seen_contacts:
                            continue
                        seen_contacts.add(signature)
                        results.append(
                            {
                                "keyword": keyword,
                                "website": site_key,
                                "title": search_result.get("title", ""),
                                "email": email,
                                "phone": phone,
                                "sourcePage": detail["pageUrl"],
                                "searchQuery": search_result.get("query", keyword),
                                "searchTitle": search_result.get("title", ""),
                                "searchSnippet": search_result.get("snippet", ""),
                            }
                        )
                        yield {
                            "type": "result_progress",
                            "websitesChecked": websites_checked,
                            "pagesChecked": pages_checked,
                            "resultsCount": len(results),
                            "latest": results[-1],
                        }
        except Exception:
            continue

    yield {
        "type": "done",
        "keyword": keyword,
        "emailFragment": email_fragment,
        "source": source,
        "searchPagesChecked": search_pages_checked,
        "websitesChecked": websites_checked,
        "pagesChecked": pages_checked,
        "searchSource": "bing-linkedin" if source == "linkedin" else "bing-python",
        "searchNotice": search_notice,
        "results": results,
    }


def main():
    payload = json.loads(sys.stdin.read() or "{}")
    keyword = str(payload.get("keyword", "")).strip()
    if not keyword:
        raise SystemExit(json.dumps({"error": "Keyword is required."}))

    data = None
    for event in scrape_leads(
        keyword=keyword,
        email_fragment=str(payload.get("emailFragment", "")).strip(),
        max_search_pages=int(payload.get("maxSearchPages", 1) or 1),
        max_websites=int(payload.get("maxWebsites", 3) or 3),
        source=str(payload.get("source", "public")).strip().lower() or "public",
    ):
        if event.get("type") == "done":
            data = event
            continue
        sys.stderr.write(json.dumps(event) + "\n")
        sys.stderr.flush()
    sys.stdout.write(json.dumps(data or {"error": "No result produced"}))


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        sys.stdout.write(json.dumps({"error": str(exc)}))
        sys.exit(1)
